src/addr.rs

```diff
@@ -1,117 +1,251 @@
-        self.0
-    }
-
-    /// Aligns the address downwards to the given alignment.
-    ///
-    /// See the [`align_down`] function for more information.
-    #[inline]
-    pub fn align_down<U>(self, align: U) -> Self
-    where
-        U: Into<usize>,
-    {
-        Self(align_down(self.0, align.into()))
-    }
-
-    /// Aligns the address upwards to the given alignment.
-    ///
-    /// See the [`align_up`] function for more information.
-    #[inline]
-    pub fn align_up<U>(self, align: U) -> Self
-    where
-        U: Into<usize>,
-    {
-        Self(align_up(self.0, align.into()))
-    }
+/// Memory address types here include both physical and virtual addresses, as well as any other
+/// similar types like guest physical addresses in a hypervisor.
+pub trait MemoryAddr:
+    // The address type should be trivially copyable. This implies `Clone`.
+    Copy
+    // The address type should be convertible to and from `usize`.
+    + From<usize>
+    + Into<usize>
+{
+    // Empty for now.
+}
 
-    /// Returns the offset of the address within the given alignment.
-    ///
-    /// See the [`align_offset`] function for more information.
-    #[inline]
-    pub fn align_offset<U>(self, align: U) -> usize
-    where
-        U: Into<usize>,
-    {
-        align_offset(self.0, align.into())
-    }
+/// Implement the `MemoryAddr` trait for any type that satisfies the required bounds.
+impl<T> MemoryAddr for T where T: Copy + From<usize> + Into<usize> {}
 
-    /// Checks whether the address has the demanded alignment.
-    ///
-    /// See the [`is_aligned`] function for more information.
-    #[inline]
-    pub fn is_aligned<U>(self, align: U) -> bool
-    where
-        U: Into<usize>,
-    {
-        is_aligned(self.0, align.into())
-    }
+/// Creates a new address type by wrapping an `usize`.
+///
+/// For each `$vis type $name;`, this macro generates the following items:
+/// - Definition of the new address type `$name`, which contains a single private unnamed field of
+///   type `usize`.
+/// - Default implementations (i.e. derived implementations) for the following traits:
+///   - `Copy`, `Clone`,
+///   - `Default`,
+///   - `Ord`, `PartialOrd`, `Eq`, and `PartialEq`.
+/// - Implementations for the following traits:
+///   - `From<usize>`, `Into<usize>` (by implementing `From<$name> for usize`),
+///   - `Add<usize>`, `AddAssign<usize>`, `Sub<usize>`, `SubAssign<usize>`, as well as
+/// - Two `const` methods to convert between the address type and `usize`:
+///   - `from_usize`, which converts an `usize` to the address type, and
+///   - `as_usize`, which converts the address type to an `usize`.
+/// - Methods to align the address, namely:
+///   - `align_down`, `align_up`, `align_offset`, `is_aligned`, `align_down_4k`, `align_up_4k`,
+///     `align_offset_4k`, and `is_aligned_4k`, which correspond to the functions with the same
+///     names in the crate root.
+#[macro_export]
+macro_rules! def_usize_addr {
```

Comment:
Add an example for the macro usage?

Reply:
Added

Reply:
Remove `fn main`

Reply:
`fn main` was added to suppress a warning in this doc test

Reply:
What error? You can also add `#` before `fn main` to hide it in rendered doc page.

Reply:
> warning: non-local `impl` definition, they should be avoided as they go against expectation

`#` is a good idea

Reply:
Fixed